src/pybind/mgr/cephadm/module.py

```diff
@@ -324,10 +333,194 @@ def __init__(self, *args, **kwargs):
             if h not in self.inventory:
                 del self.service_cache[h]
 
+        # for serve()
+        self.run = True
+        self.event = Event()
+
     def shutdown(self):
         self.log.error('shutdown')
         self._worker_pool.close()
         self._worker_pool.join()
+        self.run = False
+        self.event.set()
+
+    def _kick_serve_loop(self):
+        self.log.debug('_kick_serve_loop')
+        self.event.set()
+
+    def _do_upgrade(self, daemons):
+        # type: (List[orchestrator.ServiceDescription]) -> Optional[AsyncCompletion]
+        if not self.upgrade_state:
+            self.log.debug('_do_upgrade no state, exiting')
+            return None
+
+        target_name = self.upgrade_state.get('target_name')
+        target_id = self.upgrade_state.get('target_id')
+        target_version = self.upgrade_state.get('target_version')
+        self.log.info('Upgrade: Target is %s id %s' % (target_name, target_id))
+
+        # get all distinct container_image settings
+        image_settings = {}
+        ret, out, err = self.mon_command({
+            'prefix': 'config dump',
+            'format': 'json',
+        })
+        config = json.loads(out)
+        for opt in config:
+            if opt['name'] == 'container_image':
+                image_settings[opt['section']] = opt['value']
+
+        for daemon_type in ['mgr', 'mon', 'osd', 'rgw', 'mds']:
+            self.log.info('Upgrade: Checking %s daemons...' % daemon_type)
+            need_upgrade_self = False
+            for d in daemons:
+                if d.service_type != daemon_type:
+                    continue
+                if daemon_type == 'mgr' and \
+                   d.service_instance == self.get_mgr_id():
+                    self.log.info('Upgrade: Need to upgrade myself (mgr.%s)' %
+                                  self.get_mgr_id())
+                    need_upgrade_self = True
+                    continue
+                if d.container_image_id != target_id:
+                    while True:
+                        if daemon_type not in ['mon', 'osd', 'mds']:
+                            break
+                        ret, out, err = self.mon_command({
+                            'prefix': '%s ok-to-stop',
+                            'ids': [d.service_instance],
+                        })
+                        if err:
```

Comment:
What does `ok-to-stop` return, if the service was removed in the meantime? Right now we're waiting in an endless loop, until the service re-appears. Quick fix would be to return `None` after four tries in this inner-loop